ow_plexil/src/plexil-adapter/LanderAdapter.cpp

```diff
@@ -0,0 +1,274 @@
+  args[1].getValue(relative);
+  args[2].getValuePointer(position);
+  args[3].getValuePointer(orientation);
+  args[4].getValue(force_threshold);
+  args[5].getValue(torque_threshold);
+
+    //change real array into a vector
+  position->getContentsVector(position_vector);
+  orientation->getContentsVector(orientation_vector);
+  std::unique_ptr<CommandRecord>& cr = new_command_record(cmd, intf);
+  LanderAdapter::s_interface->armMoveCartesianGuarded (frame, relative,
+                                                       *position_vector,
+                                                       *orientation_vector,
+                                                       force_threshold,
+                                                       torque_threshold,
+                                                       CommandId);
+  acknowledge_command_sent(*cr);
+}
+
+static void arm_move_joint (Command* cmd, AdapterExecInterface* intf)
+{
+  bool relative;
+  int joint;
+  double radians;
+  const vector<Value>& args = cmd->getArgValues();
+  args[0].getValue(relative);
+  args[1].getValue(joint);
+  args[2].getValue(radians);
+  std::unique_ptr<CommandRecord>& cr = new_command_record(cmd, intf);
+  LanderAdapter::s_interface->armMoveJoint (relative, joint, radians, CommandId);
+  acknowledge_command_sent(*cr);
+}
+
+static void task_deliver_sample (Command* cmd, AdapterExecInterface* intf)
+{
+  unique_ptr<CommandRecord>& cr = new_command_record(cmd, intf);
+  LanderAdapter::s_interface->taskDeliverSample (CommandId);
+  acknowledge_command_sent(*cr);
+}
+
+static void task_discard_sample (Command* cmd, AdapterExecInterface* intf)
+{
+  int frame;
+  bool relative;
+  RealArray const* point = nullptr;
+  vector<double> const* point_vector = nullptr;
+  double height;
+
+  const vector<Value>& args = cmd->getArgValues();
+  args[0].getValue(frame);
+  args[1].getValue(relative);
+  args[2].getValuePointer(point);
+  args[3].getValue(height);
+  point->getContentsVector(point_vector);
+  unique_ptr<CommandRecord>& cr = new_command_record(cmd, intf);
+  LanderAdapter::s_interface->taskDiscardSample (frame, relative, *point_vector,
+                                                 height, CommandId);
+  acknowledge_command_sent(*cr);
+}
+
+static void pan_tilt_move_joints (Command* cmd, AdapterExecInterface* intf)
+{
+  double pan_degrees, tilt_degrees;
+  const vector<Value>& args = cmd->getArgValues();
+  args[0].getValue (pan_degrees);
+  args[1].getValue (tilt_degrees);
+  if (! LanderAdapter::checkAngle ("pan", pan_degrees, LanderAdapter::PanMinDegrees,
+                                   LanderAdapter::PanMaxDegrees,
+                                   LanderAdapter::PanTiltInputTolerance) ||
+      ! LanderAdapter::checkAngle ("tilt", tilt_degrees, LanderAdapter::TiltMinDegrees,
+                                   LanderAdapter::TiltMaxDegrees,
+                                   LanderAdapter::PanTiltInputTolerance)) {
```

Comment:
But shouldn't it be also checked at the PLEXIL level so that the command is denied outright?  If we let bad arguments pass to the action server, and it aborts, PLEXIL won't know why...

Reply:
Is Plexil not aware of the Action Result at present? If not, I think this is needed eventually because there are other ways an action could fail which Plexil won't be made aware of otherwise.

Reply:
PLEXIL has little or no awareness of action result and this is a big weakness in autonomy and a separate effort.

I think this particular check should be left here because it enforces a testbed unification policy and not necessary a technical limitation of the simulator.